File: MobileTrailer/PRCell.m
#import "PRCell.h"
#import "PullRequest.h"
#import "Settings.h"

static NSNumberFormatter *itemCountFormatter;

@implementation PRCell
{
	UILabel *unreadCount, *readCount;
	NSString *failedToLoadImage, *waitingForImageInPath;
	__weak IBOutlet UIImageView *_image;
	__weak IBOutlet UILabel *_title;
	__weak IBOutlet UILabel *_description;
}

- (void)awakeFromNib
{
	[super awakeFromNib];

	unreadCount = [[UILabel alloc] initWithFrame:CGRectZero];
	unreadCount.textColor = [COLOR_CLASS whiteColor];
	unreadCount.textAlignment = NSTextAlignmentCenter;
	unreadCount.layer.cornerRadius = 8.5;
	unreadCount.clipsToBounds = YES;
	unreadCount.font = [UIFont boldSystemFontOfSize:12.0];
	unreadCount.hidden = YES;
	[self.contentView addSubview:unreadCount];

	readCount = [[UILabel alloc] initWithFrame:CGRectZero];
	readCount.textColor = [COLOR_CLASS darkGrayColor];
	readCount.textAlignment = NSTextAlignmentCenter;
	readCount.layer.cornerRadius = 9.0;
	readCount.clipsToBounds = YES;
	readCount.font = [UIFont systemFontOfSize:12.0];
	readCount.hidden = YES;
	[self.contentView addSubview:readCount];

	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		itemCountFormatter = [[NSNumberFormatter alloc] init];
		itemCountFormatter.numberStyle = NSNumberFormatterDecimalStyle;
	});

	[[NSNotificationCenter defaultCenter] addObserver:self
											 selector:@selector(networkStateChanged)
												 name:kReachabilityChangedNotification
											   object:nil];
}

- (void)networkStateChanged
{
	if(!failedToLoadImage) return;
	if([app.api.reachability currentReachabilityStatus]!=NotReachable)
	{
		[self loadImageAtPath:failedToLoadImage];
	}
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)setPullRequest:(PullRequest *)pullRequest
{
	UIFont *detailFont = [UIFont systemFontOfSize:[UIFont smallSystemFontSize]];

	_title.attributedText = [pullRequest titleWithFont:_title.font
											 labelFont:[detailFont fontWithSize:detailFont.pointSize-2.0]
											titleColor:[COLOR_CLASS darkTextColor]];

	_description.attributedText = [pullRequest subtitleWithFont:detailFont
													 lightColor:[COLOR_CLASS lightGrayColor]
													  darkColor:[COLOR_CLASS darkGrayColor]];

	NSInteger _commentsNew=0;
	NSInteger _commentsTotal = pullRequest.totalComments.integerValue;
	if(settings.showCommentsEverywhere || pullRequest.isMine || pullRequest.commentedByMe)
	{
		_commentsNew = pullRequest.unreadComments.integerValue;
	}

	readCount.text = [itemCountFormatter stringFromNumber:@(_commentsTotal)];
	CGSize size = [readCount sizeThatFits:CGSizeMake(200, 14.0)];
	readCount.frame = CGRectMake(0, 0, size.width+10.0, 17.0);
	readCount.hidden = (_commentsTotal==0);

	unreadCount.hidden = _commentsNew==0;
	unreadCount.text = [itemCountFormatter stringFromNumber:@(_commentsNew)];
	size = [unreadCount sizeThatFits:CGSizeMake(200, 18.0)];
	unreadCount.frame = CGRectMake(0, 0, size.width+10.0, 17.0);

	NSString *imagePath = pullRequest.userAvatarUrl;
	if(imagePath)
		[self loadImageAtPath:imagePath];
	else
		failedToLoadImage = nil;

	self.accessibilityLabel = [NSString stringWithFormat:@"%@, %@ unread comments, %@ total comments, %@",
							   [pullRequest accessibleTitle],
							   unreadCount.text,
							   readCount.text,
							   [pullRequest accessibleSubtitle]];

	[self setNeedsLayout];
}

- (void)loadImageAtPath:(NSString *)imagePath
{
	waitingForImageInPath = imagePath;
	if(![app.api haveCachedAvatar:imagePath
			   tryLoadAndCallback:^(id image) {
				   if([waitingForImageInPath isEqualToString:imagePath])
				   {
					   if(image)
					   {
						   // image loaded
						   _image.image = image;
						   failedToLoadImage = nil;
					   }
					   else
					   {
						   // load failed / no image
						   _image.image = [UIImage imageNamed:@"avatarPlaceHolder"];
						   failedToLoadImage = imagePath;
					   }
					   waitingForImageInPath = nil;
				   }
			   }])
	{
		// prepare UI for over-the-network load
		_image.image = [UIImage imageNamed:@"avatarPlaceHolder"];
		failedToLoadImage = nil;
	}
}

- (void)layoutSubviews
{
	[super layoutSubviews];

	CGPoint topLeft = CGPointMake(_image.frame.origin.x, _image.frame.origin.y);
	unreadCount.center = topLeft;
	[self.contentView bringSubviewToFront:unreadCount];

	CGPoint bottomRight = CGPointMake(topLeft.x+_image.frame.size.width, topLeft.y+_image.frame.size.height);
	readCount.center = bottomRight;
	[self.contentView bringSubviewToFront:readCount];
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated
{
	[super setSelected:selected animated:animated];
	[self tone:selected];
}

- (void)setHighlighted:(BOOL)highlighted animated:(BOOL)animated
{
	[super setHighlighted:highlighted animated:animated];
	[self tone:highlighted];
}

- (void)tone:(BOOL)tone
{
	unreadCount.backgroundColor = [COLOR_CLASS redColor];
	readCount.backgroundColor = [COLOR_CLASS colorWithWhite:0.9 alpha:1.0];
}

@end

File: MobileTrailer/PRCell.swift

import UIKit

class PRCell: UITableViewCell {

	private let unreadCount = CountLabel(frame: CGRectZero)
	private let readCount = CountLabel(frame: CGRectZero)
	private var failedToLoadImage: NSString?
	private var waitingForImageInPath: NSString?

	@IBOutlet weak var _image: UIImageView!
	@IBOutlet weak var _title: UILabel!
	@IBOutlet weak var _description: UILabel!
	@IBOutlet weak var _statuses: UILabel!

	override func awakeFromNib() {
		unreadCount.textColor = UIColor.whiteColor()
		contentView.addSubview(unreadCount)

		readCount.textColor = UIColor.darkGrayColor()
		contentView.addSubview(readCount)

		NSNotificationCenter.defaultCenter().addObserver(self, selector: Selector("networkStateChanged"), name: kReachabilityChangedNotification, object: nil)
	}

	private var needsConstraints = true

	override func updateConstraints() {

		if needsConstraints {
			needsConstraints = false

			contentView.addConstraints([

				NSLayoutConstraint(item: _image,
					attribute: NSLayoutAttribute.CenterX,
					relatedBy: NSLayoutRelation.Equal,
					toItem: readCount,
					attribute: NSLayoutAttribute.CenterX,
					multiplier: 1,
					constant: -23),

				NSLayoutConstraint(item: _image,
					attribute: NSLayoutAttribute.CenterY,
					relatedBy: NSLayoutRelation.Equal,
					toItem: readCount,
					attribute: NSLayoutAttribute.CenterY,
					multiplier: 1,
					constant: -23),

				NSLayoutConstraint(item: _image,
					attribute: NSLayoutAttribute.CenterX,
					relatedBy: NSLayoutRelation.Equal,
					toItem: unreadCount,
					attribute: NSLayoutAttribute.CenterX,
					multiplier: 1,
					constant: 23),

				NSLayoutConstraint(item: _image,
					attribute: NSLayoutAttribute.CenterY,
					relatedBy: NSLayoutRelation.Equal,
					toItem: unreadCount,
					attribute: NSLayoutAttribute.CenterY,
					multiplier: 1,
					constant: 23)
				])
		}
		super.updateConstraints()
	}

	func networkStateChanged() {
		dispatch_async(dispatch_get_main_queue()) { [weak self] in
			if self!.failedToLoadImage == nil { return }
			if api.currentNetworkStatus != NetworkStatus.NotReachable {
				self!.loadImageAtPath(self!.failedToLoadImage)
			}
		}
	}

	deinit {
		NSNotificationCenter.defaultCenter().removeObserver(self)
	}

	func setPullRequest(pullRequest: PullRequest) {
		let detailFont = UIFont.systemFontOfSize(UIFont.smallSystemFontSize())
		_title.attributedText = pullRequest.titleWithFont(_title.font, labelFont: detailFont.fontWithSize(detailFont.pointSize-2), titleColor: UIColor.darkTextColor())
		_description.attributedText = pullRequest.subtitleWithFont(detailFont, lightColor: UIColor.lightGrayColor(), darkColor: UIColor.darkGrayColor())
		var statusText : NSMutableAttributedString?
		var statusCount = 0
		if Settings.showStatusItems {
			let statusItems = pullRequest.displayedStatuses() + pullRequest.displayedStatuses()
			statusCount = statusItems.count
			if statusCount > 0 {

				let paragraphStyle = NSMutableParagraphStyle()
				paragraphStyle.headIndent = 103.0
				paragraphStyle.paragraphSpacing = 6.0

				let statusAttributes = [
					NSFontAttributeName: UIFont(name: "Courier", size: 10)!,
					NSParagraphStyleAttributeName: paragraphStyle]

				statusText = NSMutableAttributedString()
				statusText?.appendAttributedString(NSAttributedString(string: "\n", attributes: statusAttributes))

				for status in statusItems {
					var lineAttributes = statusAttributes
					lineAttributes[NSForegroundColorAttributeName] = status.colorForDisplay()
					let text = status.displayText()
					statusText?.appendAttributedString(NSAttributedString(string: text, attributes: lineAttributes))
					if --statusCount > 0 {
						statusText?.appendAttributedString(NSAttributedString(string: "\n", attributes: lineAttributes))
					}
				}
			}
		}
		_statuses.attributedText = statusText

		var _commentsNew = 0
		let _commentsTotal = pullRequest.totalComments?.integerValue ?? 0

		if Settings.showCommentsEverywhere || pullRequest.isMine() || pullRequest.commentedByMe() {
			_commentsNew = pullRequest.unreadComments?.integerValue ?? 0
		}

		readCount.text = itemCountFormatter.stringFromNumber(_commentsTotal)
		let readSize = readCount.sizeThatFits(CGSizeMake(200, 14))
		readCount.hidden = (_commentsTotal == 0)

		unreadCount.hidden = (_commentsNew == 0)
		unreadCount.text = itemCountFormatter.stringFromNumber(_commentsNew)

		loadImageAtPath(pullRequest.userAvatarUrl)

		if let statusString = statusText?.string {
			accessibilityLabel = "\(pullRequest.accessibleTitle()), \(unreadCount.text) unread comments, \(readCount.text) total comments, \(pullRequest.accessibleSubtitle()). \(statusCount) statuses: \(statusString)"
		} else {
			accessibilityLabel = "\(pullRequest.accessibleTitle()), \(unreadCount.text) unread comments, \(readCount.text) total comments, \(pullRequest.accessibleSubtitle())"
		}

		setNeedsLayout()
	}

	private func loadImageAtPath(imagePath: NSString?) {
		waitingForImageInPath = imagePath
		if let path = imagePath {
			if !api.haveCachedAvatar(path, tryLoadAndCallback: { [weak self] image in
				if self!.waitingForImageInPath == path {
					if image != nil {
						// image loaded
						self!._image.image = image
						self!.failedToLoadImage = nil
					} else {
						// load failed / no image
						self!._image.image = UIImage(named: "avatarPlaceHolder")
						self!.failedToLoadImage = imagePath
					}
					self!.waitingForImageInPath = nil
				}
			}) {
				// prepare UI for over-the-network load
				_image.image = UIImage(named: "avatarPlaceHolder")
				failedToLoadImage = nil
			}
		} else {
			failedToLoadImage = nil
		}
	}

	override func layoutSubviews() {
		super.layoutSubviews()

		_title.preferredMaxLayoutWidth = _title.bounds.size.width
		_description.preferredMaxLayoutWidth = _description.bounds.size.width
		_statuses.preferredMaxLayoutWidth = _statuses.bounds.size.width
	}

	override func setSelected(selected: Bool, animated: Bool) {
		super.setSelected(selected, animated:animated)
		tone(selected)
	}

	override func setHighlighted(highlighted: Bool, animated: Bool) {
		super.setHighlighted(highlighted, animated:animated)
		tone(highlighted)
	}

	func tone(tone: Bool)
	{
		unreadCount.backgroundColor = UIColor.redColor()
		readCount.backgroundColor = UIColor(white: 0.9, alpha: 1.0)
	}
}